src/Compilers/Core/Portable/SourceGeneration/Nodes/IncrementalValueSources.cs

```diff
@@ -40,46 +47,86 @@ internal static class SharedInputNodes
     internal sealed class PerGeneratorInputNodes
     {
-        public static PerGeneratorInputNodes Empty = new PerGeneratorInputNodes();
-
-        private PerGeneratorInputNodes() { }
+        public static readonly PerGeneratorInputNodes Empty = new PerGeneratorInputNodes(receiverNode: null, ImmutableArray<ISyntaxTransformNode>.Empty);
 
-        private PerGeneratorInputNodes(InputNode<ISyntaxContextReceiver>? receiverNode)
+        private PerGeneratorInputNodes(InputNode<ISyntaxContextReceiver>? receiverNode, ImmutableArray<ISyntaxTransformNode> transformNodes)
         {
             this.ReceiverNode = receiverNode;
+            this.TransformNodes = transformNodes;
         }
 
+        public ImmutableArray<ISyntaxTransformNode> TransformNodes { get; }
+
         public InputNode<ISyntaxContextReceiver>? ReceiverNode { get; }
 
         public sealed class Builder
         {
             private InputNode<ISyntaxContextReceiver>? _receiverNode;
 
-            bool disposed = false;
+            private ArrayBuilder<ISyntaxTransformNode>? _transformNodes;
+
+            bool _disposed = false;
 
-            public Builder()
+            public InputNode<ISyntaxContextReceiver> GetOrCreateReceiverNode()
             {
+                Debug.Assert(!_disposed);
+                if (_receiverNode is null)
+                {
+                    // this is called by only internal code which we know to be thread safe
+                    _receiverNode = new InputNode<ISyntaxContextReceiver>();
+                }
+                return _receiverNode;
             }
 
-            public InputNode<ISyntaxContextReceiver> GetOrCreateReceiverNode()
+            public ArrayBuilder<ISyntaxTransformNode> SyntaxTransformNodes
             {
-                Debug.Assert(!disposed);
-                return InterlockedOperations.Initialize(ref _receiverNode, new InputNode<ISyntaxContextReceiver>());
+                get
+                {
+                    Debug.Assert(!_disposed);
+                    if (_transformNodes is null)
+                    {
+                        // PROTOTYPE(source-generators): this is resilient to threading in the user pipeline
+                        // but the rest of the structure isn't. We should decide if we want to support that or 
+                        // just say this whole thing isn't thread safe.
+
+                        var newNodes = ArrayBuilder<ISyntaxTransformNode>.GetInstance();
+                        var storedNodes = InterlockedOperations.Initialize(ref _transformNodes, newNodes);
+
+                        // in the case another thread beat us to intialization, we will see that threads arraybuilder.
+                        // free the one we created and pass back the other threads
+                        if (newNodes != storedNodes)
+                        {
+                            newNodes.Free();
+                        }
+                        return storedNodes;
```

Comment:
>return storedNodes; [](http://example.com/codeflow?start=24&length=19)

Minor: This `return` is not necessary, and we could compare `_transformNodes` above and remove the `storedNodes` local.